### ISSUE TITLE
Declaration order not preserved! Compilation fails!!

### ISSUE THREAD

**Issue by clementval** [omni-compiler/omni-compiler#264]
```fortran
MODULE mod1

CONTAINS
  SUBROUTINE exchg_boundaries (kdim, var01)
    IMPLICIT NONE

    INTEGER (KIND=4), INTENT(IN) :: kdim(24)
    REAL (KIND=8), INTENT (INOUT) :: var01(kdim(1))

    TYPE :: pointerto3d
      REAL(KIND=8) :: p
    END TYPE pointerto3d

    IF (kdim(1) <= 0) THEN
    END IF

  END SUBROUTINE exchg_boundaries
END MODULE mod1
```

After transformation, the code will be like the following: 

```fortran
MODULE mod1

CONTAINS
 SUBROUTINE exchg_boundaries ( kdim , var01 )
  REAL ( KIND= 8 ) , INTENT(INOUT) :: var01 ( 1 : kdim ( 1 ) )
  TYPE :: pointerto3d
   REAL ( KIND= 8 ) :: p
  END TYPE pointerto3d
  INTEGER ( KIND= 4 ) , INTENT(IN) :: kdim ( 1 : 24 )


  IF ( kdim ( 1 ) <= 0 ) THEN
  END IF
 END SUBROUTINE exchg_boundaries

END MODULE mod1 
```

The `INTEGER` `kdim` is not declared at the right place. If I remove the type declaration, the problem disappear. 

If we have a look at the XcodeML/F, the line of the `kdim` declaration is also wrong and report to the line when it is used in the conditional.
```xml
<XcodeProgram source="./F-FrontEnd/test/testdata/issue264.f90"
              language="Fortran"
              time="2017-09-11 20:33:05"
              compiler-info="XcodeML/Fortran-FrontEnd"
              version="1.0">
  <typeTable>
    <FfunctionType type="F7fc544408130" return_type="Fvoid">
      <params>
        <name type="A7fc544409780">kdim</name>
        <name type="A7fc54440a6e0">var01</name>
      </params>
    </FfunctionType>
    <FbasicType type="I7fc544409540" ref="Fint">
      <kind>
        <FintConstant type="Fint">4</FintConstant>
      </kind>
    </FbasicType>
    <FbasicType type="I7fc544409660" intent="in" ref="I7fc544409540"/>
    <FbasicType type="A7fc544409780" intent="in" ref="I7fc544409660">
      <indexRange>
        <lowerBound>
          <FintConstant type="Fint">1</FintConstant>
        </lowerBound>
        <upperBound>
          <FintConstant type="Fint">24</FintConstant>
        </upperBound>
      </indexRange>
    </FbasicType>
    <FbasicType type="R7fc54440a4a0" ref="Freal">
      <kind>
        <FintConstant type="Fint">8</FintConstant>
      </kind>
    </FbasicType>
    <FbasicType type="R7fc54440a5c0" intent="inout" ref="R7fc54440a4a0"/>
    <FbasicType type="A7fc54440a6e0" intent="inout" ref="R7fc54440a5c0">
      <indexRange>
        <lowerBound>
          <FintConstant type="Fint">1</FintConstant>
        </lowerBound>
        <upperBound>
          <FarrayRef type="I7fc544409660">
            <varRef type="A7fc544409780">
              <Var type="A7fc544409780" scope="local">kdim</Var>
            </varRef>
            <arrayIndex>
              <FintConstant type="Fint">1</FintConstant>
            </arrayIndex>
          </FarrayRef>
        </upperBound>
      </indexRange>
    </FbasicType>
    <FstructType type="S7fc54440aa60">
      <symbols>
        <id type="R7fc54440b4e0">
          <name>p</name>
        </id>
      </symbols>
    </FstructType>
    <FbasicType type="R7fc54440b2e0" ref="Freal">
      <kind>
        <FintConstant type="Fint">8</FintConstant>
      </kind>
    </FbasicType>
    <FbasicType type="R7fc54440b4e0" ref="R7fc54440b2e0"/>
  </typeTable>
  <globalSymbols>
    <id sclass="ffunc">
      <name>mod1</name>
    </id>
  </globalSymbols>
  <globalDeclarations>
    <FmoduleDefinition name="mod1" lineno="1" file="./F-FrontEnd/test/testdata/issue264.f90">
      <symbols>
        <id type="F7fc544408130" sclass="ffunc">
          <name>exchg_boundaries</name>
        </id>
      </symbols>
      <declarations>
      </declarations>
      <FcontainsStatement lineno="3" file="./F-FrontEnd/test/testdata/issue264.f90">
        <FfunctionDefinition lineno="4" file="./F-FrontEnd/test/testdata/issue264.f90">
          <name type="F7fc544408130">exchg_boundaries</name>
          <symbols>
            <id type="F7fc544408130" sclass="ffunc">
              <name>exchg_boundaries</name>
            </id>
            <id type="A7fc544409780" sclass="fparam">
              <name>kdim</name>
            </id>
            <id type="A7fc54440a6e0" sclass="fparam">
              <name>var01</name>
            </id>
            <id type="S7fc54440aa60" sclass="ftype_name">
              <name>pointerto3d</name>
            </id>
          </symbols>
          <declarations>
            <varDecl lineno="4" file="./F-FrontEnd/test/testdata/issue264.f90">
              <name type="F7fc544408130">exchg_boundaries</name>
            </varDecl>
            <varDecl lineno="8" file="./F-FrontEnd/test/testdata/issue264.f90">
              <name type="A7fc54440a6e0">var01</name>
            </varDecl>
            <FstructDecl lineno="10" file="./F-FrontEnd/test/testdata/issue264.f90">
              <name type="S7fc54440aa60">pointerto3d</name>
            </FstructDecl>
            <varDecl lineno="14" file="./F-FrontEnd/test/testdata/issue264.f90">
              <name type="A7fc544409780">kdim</name>
            </varDecl>
          </declarations>
          <body>
            <FifStatement lineno="14" file="./F-FrontEnd/test/testdata/issue264.f90">
              <condition>
                <logLEExpr type="Flogical">
                  <FarrayRef type="I7fc544409660">
                    <varRef type="A7fc544409780">
                      <Var type="A7fc544409780" scope="local">kdim</Var>
                    </varRef>
                    <arrayIndex>
                      <FintConstant type="Fint">1</FintConstant>
                    </arrayIndex>
                  </FarrayRef>
                  <FintConstant type="Fint">0</FintConstant>
                </logLEExpr>
              </condition>
              <then>
                <body>
                </body>
              </then>
            </FifStatement>
          </body>
        </FfunctionDefinition>
      </FcontainsStatement>
    </FmoduleDefinition>
  </globalDeclarations>
</XcodeProgram>
```

**Comment by clementval** [comment]
PR #265 ready for merge
